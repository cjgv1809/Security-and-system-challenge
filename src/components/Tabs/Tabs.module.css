.tabs {
  display: flex;
  flex-direction: column;
  width: 100%;
  padding: 2rem;
}

.tabList {
  list-style: none;
  margin: 0 auto 1rem;
  padding: 0.25rem;
  display: flex;
  gap: 0.25rem;
  justify-content: center;
  align-items: center;
  background-color: var(--color-primary);
  width: fit-content;
  border-radius: 10px;
  box-shadow: 0 -1px 20px var(--color-box-shadow);
}

.tab,
.activeTab {
  padding: 0.8rem 3rem;
  border: none;
  border-radius: 6px;
  cursor: pointer;
  white-space: nowrap;
  background-color: var(--color-primary);
  color: var(--color-secondary);
  transition: background-color 0.3s ease-in-out, color 0.3s ease-in-out;
}

.activeTab {
  background-color: var(--color-active-background);
  color: var(--color-active-text);
}

.content {
  padding-top: 1.5rem;
  margin: 0 auto;
  border-top: 1px solid var(--color-content-border);
  width: min(100%, 800px);
}

.activeContent {
  display: block;
}

.hiddenContent {
  display: none;
}

@media screen and (max-width: 600px) {
  .tabs {
    padding: 1rem;
  }

  .tabList {
    gap: 0.3125rem;
  }

  .tab,
  .activeTab {
    padding: 0.3125rem 0.625rem;
  }
}
