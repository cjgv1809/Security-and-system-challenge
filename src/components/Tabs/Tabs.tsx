"use client";

import React, { useState, useRef, useEffect } from "react";
import { motion } from "framer-motion";
import styles from "./Tabs.module.css";

type TabProps = {
  title: string;
  children: React.ReactNode;
};

type TabsProps = {
  defaultIndex?: number; // Optional default active tab index
  children: React.ReactElement<TabProps>[]; // Tabs as children
};

const Tabs: React.FC<TabsProps> = ({ defaultIndex = 0, children }) => {
  const [activeTabIndex, setActiveTabIndex] = useState(defaultIndex);
  const tabRefs = useRef<(HTMLButtonElement | null)[]>([]);

  useEffect(() => {
    tabRefs.current[activeTabIndex]?.focus();
  }, [activeTabIndex]);

  const handleClick = (index: number) => {
    setActiveTabIndex(index);
  };

  return (
    <div className={styles.tabs}>
      <ul className={styles.tabList} role="tablist" aria-label="PestaÃ±as">
        {children.map((child, index) => (
          <li key={index} className={styles.tabItem}>
            <motion.button
              type="button"
              className={`${styles.tab} ${
                activeTabIndex === index ? styles.activeTab : ""
              }`}
              onClick={() => handleClick(index)}
              role="tab"
              id={`tab-${index}`}
              aria-selected={activeTabIndex === index}
              aria-controls={`tabpanel-${index}`}
              ref={(el) => {
                tabRefs.current[index] = el;
              }}
              initial={{ opacity: 0, x: -20 }}
              animate={{ opacity: 1, x: 0 }}
              transition={{ duration: 0.3 }}
              exit={{ opacity: 0, x: 20 }}
            >
              {child.props.title}
            </motion.button>
          </li>
        ))}
      </ul>
      <div className={styles.content}>
        {children.map((child, index) => {
          return (
            activeTabIndex === index && (
              <motion.div
                key={index}
                className={styles.activeContent}
                role="tabpanel"
                id={`tabpanel-${index}`}
                aria-labelledby={`tab-${index}`}
                hidden={activeTabIndex !== index}
                initial={{ opacity: 0, y: -40 }}
                animate={{ opacity: 1, y: 0 }}
                transition={{ duration: 0.5, delay: 0.05 }}
              >
                {child.props.children}
              </motion.div>
            )
          );
        })}
      </div>
    </div>
  );
};

export const Tab: React.FC<TabProps> = ({ children }) => <>{children}</>;

export default Tabs;
