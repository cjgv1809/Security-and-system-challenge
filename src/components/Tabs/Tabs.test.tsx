import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Tabs, { Tab } from "./Tabs";

describe("Tabs Component", () => {
  it("should render the Tabs component", () => {
    render(
      <Tabs>
        <Tab title="Tab 1">Content 1</Tab>
        <Tab title="Tab 2">Content 2</Tab>
        <Tab title="Tab 3">Content 3</Tab>
      </Tabs>
    );

    expect(screen.getByText("Tab 1")).toBeInTheDocument();
    expect(screen.getByText("Tab 2")).toBeInTheDocument();
    expect(screen.getByText("Tab 3")).toBeInTheDocument();
    expect(screen.getByText("Content 1")).toBeInTheDocument();
    expect(screen.queryByText("Content 2")).not.toBeInTheDocument();
    expect(screen.queryByText("Content 3")).not.toBeInTheDocument();
  });

  it("should change tabs on click", () => {
    render(
      <Tabs>
        <Tab title="Tab 1">Content 1</Tab>
        <Tab title="Tab 2">Content 2</Tab>
        <Tab title="Tab 3">Content 3</Tab>
      </Tabs>
    );

    fireEvent.click(screen.getByText("Tab 2"));
    expect(screen.getByText("Content 2")).toBeInTheDocument();
    expect(screen.queryByText("Content 1")).not.toBeInTheDocument();
    expect(screen.queryByText("Content 3")).not.toBeInTheDocument();

    fireEvent.click(screen.getByText("Tab 3"));
    expect(screen.getByText("Content 3")).toBeInTheDocument();
    expect(screen.queryByText("Content 1")).not.toBeInTheDocument();
    expect(screen.queryByText("Content 2")).not.toBeInTheDocument();
  });

  it("should focus the active tab", () => {
    render(
      <Tabs>
        <Tab title="Tab 1">Content 1</Tab>
        <Tab title="Tab 2">Content 2</Tab>
        <Tab title="Tab 3">Content 3</Tab>
      </Tabs>
    );

    const tab2 = screen.getByText("Tab 2");
    fireEvent.click(tab2);

    expect(tab2).toHaveFocus();
  });

  it("should match the snapshot", () => {
    const { container } = render(
      <Tabs>
        <Tab title="Tab 1">Content 1</Tab>
        <Tab title="Tab 2">Content 2</Tab>
        <Tab title="Tab 3">Content 3</Tab>
      </Tabs>
    );

    expect(container).toMatchSnapshot();
  });
});
