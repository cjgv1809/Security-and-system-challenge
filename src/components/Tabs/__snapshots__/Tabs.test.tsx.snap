// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component should match the snapshot 1`] = `
<div>
  <div
    class="tabs"
  >
    <ul
      aria-label="PestaÃ±as"
      class="tabList"
      role="tablist"
    >
      <li
        class="tabItem"
      >
        <button
          aria-controls="tabpanel-0"
          aria-selected="true"
          class="tab activeTab"
          id="tab-0"
          role="tab"
          style="opacity: 0; transform: translateX(-20px) translateZ(0);"
          type="button"
        >
          Tab 1
        </button>
      </li>
      <li
        class="tabItem"
      >
        <button
          aria-controls="tabpanel-1"
          aria-selected="false"
          class="tab "
          id="tab-1"
          role="tab"
          style="opacity: 0; transform: translateX(-20px) translateZ(0);"
          type="button"
        >
          Tab 2
        </button>
      </li>
      <li
        class="tabItem"
      >
        <button
          aria-controls="tabpanel-2"
          aria-selected="false"
          class="tab "
          id="tab-2"
          role="tab"
          style="opacity: 0; transform: translateX(-20px) translateZ(0);"
          type="button"
        >
          Tab 3
        </button>
      </li>
    </ul>
    <div
      class="content"
    >
      <div
        aria-labelledby="tab-0"
        class="activeContent"
        id="tabpanel-0"
        role="tabpanel"
        style="opacity: 0; transform: translateY(-40px) translateZ(0);"
      >
        Content 1
      </div>
    </div>
  </div>
</div>
`;
